### PR TITLE
bpo-30871: Add test.pythoninfo

### PR DESCRIPTION
**Pull request by vstinner**
Add Lib/test/pythoninfo.py: script collecting various informations
about Python to help debugging test failures.

https://bugs.python.org/issue30871

### CODE REVIEW

**Comment by vstinner** [comment]
@zware: Would you mind to check if I added the pythoninfo run to all ways to run Python tests on all our CIs?

**Comment by vstinner** [comment]
Oh, AppVeyor doesn't want to run on my PR? I'm not sure about my PCbuild/rt.bat change.

**Comment by zware** [comment]
If you're only going for CI runs, then it should be removed from `PCbuild\rt.bat` and added to `Tools\buildbot\test.bat` instead.  If you're going for all test runs, it could be removed from `make buildbottest` and added to `Tools\scripts\run_test.py` instead.

I think my actual preference would be to add a CLI option to regrtest, and add that flag to `make buildbottest` and `Tools\buildbot\test.bat`.

Also, AppVeyor isn't running because the most recent commit only changes .travis.yml, which AppVeyor ignores.

File: Lib/test/pythoninfo.py [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+class PythonInfo:
+    def __init__(self):
+        self.info = {}
+
+    def add(self, key, value):
+        if key in self.info:
+            raise ValueError("duplicate key: %r" % key)
+
+        # accepted types: str, tuple of str, None
```
Comment (JulienPalard) [inline Lib/test/pythoninfo.py]: tuples of str does not looks accepted here, the following TypeError looks more in-sync with the implementation "value type must be str, int or None". Also I'd place this kind of comment in the docstring, it's not that far and may be usefull?

File: Lib/test/pythoninfo.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        elif value is None:
+            return
+        elif not isinstance(value, INT_TYPES):
+            raise TypeError("value type must be str, int or None")
+
+        self.info[key] = value
+
+    def get_infos(self):
+        """
+        Get informations at a key:value dictionary where values are strings.
```
Comment (JulienPalard) [inline Lib/test/pythoninfo.py]: s/at/as/

File: Lib/test/pythoninfo.py (hunk 1 above)
Comment (vstinner) [reply Lib/test/pythoninfo.py]: comment removed

File: Lib/test/pythoninfo.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/pythoninfo.py]: fixed

**Comment by vstinner** [comment]
@zware: "If you're only going for CI runs, then it should be removed from PCbuild\rt.bat and added to Tools\buildbot\test.bat instead. If you're going for all test runs, it could be removed from make buildbottest and added to Tools\scripts\run_test.py instead. I think my actual preference would be to add a CLI option to regrtest, and add that flag to make buildbottest and Tools\buildbot\test.bat."

I don't want to pollute test output with pythoninfo. My plan is to run pythoninfo as a separated command to be able to collapse its output.

In AppVeyor and Travis CI config, I added the pythoninfo run in the build step, so it may be collapsed by default.

"Also, AppVeyor isn't running because the most recent commit only changes .travis.yml, which AppVeyor ignores."

Oh ok. I rebased my patch serie into a single commit and rebased it, so AppVeyor will run.

**Comment by vstinner** [comment]
@zware: Oh, I forgot to mention that I plan to add a new "pythoninfo" step on buildbots once pythoninfo is added to 2.7, 3.6 and master branches.

**Review by zware — APPROVED** [review]
I've not reviewed the script itself in detail, but the CI changes LGTM.

**Pull request merged**